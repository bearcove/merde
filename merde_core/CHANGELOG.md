# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [4.0.1](https://github.com/bearcove/merde/compare/merde_core-v4.0.0...merde_core-v4.0.1) - 2024-09-14

### Other

- Add serde feature for merde/merde_core for CowStr

## [3.0.1](https://github.com/bearcove/merde/compare/merde_core-v3.0.0...merde_core-v3.0.1) - 2024-09-12

### Other

- Remove unused dependencies

## [2.2.3](https://github.com/bearcove/merde_json/compare/merde_json_types-v2.2.2...merde_json_types-v2.2.3) - 2024-09-05

### Other
- Update logo attribution

## [2.2.2](https://github.com/bearcove/merde_json/compare/merde_json_types-v2.2.1...merde_json_types-v2.2.2) - 2024-08-16

### Other
- updated the following local packages: merde_json, merde_json

## [2.2.1](https://github.com/bearcove/merde_json/compare/merde_json_types-v2.2.0...merde_json_types-v2.2.1) - 2024-08-16

### Other
- updated the following local packages: merde_json, merde_json

## [2.2.0](https://github.com/bearcove/merde_json/compare/merde_json_types-v2.1.2...merde_json_types-v2.2.0) - 2024-08-16

### Added
- Provide Fantome from both merde-json and merde-json-types

## [2.1.2](https://github.com/bearcove/merde_json/compare/merde_json_types-v2.1.1...merde_json_types-v2.1.2) - 2024-08-16

### Other
- updated the following local packages: merde_json, merde_json

## [2.1.1](https://github.com/bearcove/merde_json/compare/merde_json_types-v2.1.0...merde_json_types-v2.1.1) - 2024-07-31

### Other
- Use public URL, hopefully works on crates too?
- Add @2x asset
- Add logo

## [2.1.0](https://github.com/bearcove/merde_json/compare/merde_json_types-v2.0.0...merde_json_types-v2.1.0) - 2024-07-31

### Added
- Add From<T> for Rfc3339 wrapper

## [2.0.0](https://github.com/bearcove/merde_json/releases/tag/merde_json_types-v2.0.0) - 2024-07-31

### Added
- Introduce merde_json_types
