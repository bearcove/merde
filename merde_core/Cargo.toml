[package]
edition = "2021"
name = "merde_core"
version = "7.0.0"
authors = ["Amos Wenger <amos@bearcove.net>"]
description = "Base types for merde"
license = "Apache-2.0 OR MIT"
readme = "README.md"
repository = "https://github.com/bearcove/merde"
keywords = ["merde", "serialization", "deserialization"]
categories = ["encoding", "parser-implementations"]
rust-version = "1.75"

[dependencies]
compact_str = { version = "0.8.0", optional = true }
compact_bytes = { version = "0.1.3", optional = true }
ordered-float = "4.3.0"
rubicon = "3.4.9"
rusqlite = { version = "0.32.1", optional = true }
serde = { version = "1", optional = true }

[features]
default = []
full = [
    # (1 per line)
    "compact_str",
    "compact_bytes",
    "serde",
    "rusqlite",
]
compact_str = ["dep:compact_str"]
compact_bytes = ["dep:compact_bytes"]
serde = ["dep:serde", "compact_str/serde"]
rusqlite = ["dep:rusqlite"]

[dev-dependencies]
trybuild = "1.0.101"
