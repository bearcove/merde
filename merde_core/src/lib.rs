mod cowstr;
pub use cowstr::CowStr;

mod array;
pub use array::Array;

mod map;
pub use map::Map;

mod error;
pub use error::MerdeError;
pub use error::ValueType;

mod into_static;
pub use into_static::IntoStatic;

mod with_lifetime;
pub use with_lifetime::WithLifetime;

mod value;
pub use value::Value;

mod deserialize;
pub use deserialize::ValueDeserialize;

/// Interpret a &[`Value`] as an instance of type `T`. This may involve
/// more cloning than [`from_value`].
pub fn from_value_ref<'s, T>(value: &Value<'s>) -> Result<T, MerdeError>
where
    T: ValueDeserialize<'s>,
{
    T::from_value_ref(Some(value))
}

/// Interpret a [`Value`] as an instance of type `T`.
pub fn from_value<'s, T>(value: Value<'s>) -> Result<T, MerdeError>
where
    T: ValueDeserialize<'s>,
{
    T::from_value(Some(value))
}
