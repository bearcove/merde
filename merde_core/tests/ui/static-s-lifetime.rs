use merde_core::FieldSlot;

fn main() {
    let mut option: Option<i32> = None;
    let slot = FieldSlot::new(&mut option);

    #[allow(clippy::needless_lifetimes)]
    fn take_static_fieldslot<'borrow>(_f: FieldSlot<'static, 'borrow>) {}

    take_static_fieldslot(slot);
}
