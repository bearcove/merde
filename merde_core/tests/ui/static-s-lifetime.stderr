error[E0597]: `option` does not live long enough
  --> tests/ui/static-s-lifetime.rs:5:31
   |
4  |     let mut option: Option<i32> = None;
   |         ---------- binding `option` declared here
5  |     let slot = FieldSlot::new(&mut option);
   |                ---------------^^^^^^^^^^^-
   |                |              |
   |                |              borrowed value does not live long enough
   |                argument requires that `option` is borrowed for `'static`
...
11 | }
   | - `option` dropped here while still borrowed
